velox/row/UnsafeRowFast.h

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/vector/ComplexVector.h"
+#include "velox/vector/DecodedVector.h"
+
+namespace facebook::velox::row {
+
+class UnsafeRowFast {
+ public:
+  UnsafeRowFast(const RowVectorPtr& vector);
+
+  /// Return row size if all fields are fixed width. Return std::nullopt if
+  /// there are variable-width fields.
+  static std::optional<int32_t> fixedRowSize(const RowTypePtr& rowType);
+
+  /// Return serialized size of the row at specified index. Use only if
+  /// 'fixedRowSize' returned std::nullopt.
+  int32_t rowSize(vector_size_t index);
+
+  /// Serializes row at specified index into 'buffer'.
+  /// 'buffer' must have sufficient capacity and set to all zeros.
+  int32_t serialize(vector_size_t index, char* buffer);
+
+ protected:
+  UnsafeRowFast(const VectorPtr& vector);
+
+  void initialize(const TypePtr& type);
+
+  bool isNullAt(vector_size_t);
+
+  /// Fixed-width types only. Returns number of bytes used by single value.
+  int32_t valueBytes() const {
+    return valueBytes_;
+  }
+
+  /// Writes fixed-width value at specified index into 'buffer'. Value must not
+  /// be null.
+  void serializeFixedWidth(vector_size_t index, char* buffer);
+
+  void
+  serializeFixedWidth(vector_size_t offset, vector_size_t size, char* buffer);
+
+  /// Returns serialized size of variable-width row.
+  int32_t variableWidthRowSize(vector_size_t index);
+
+  /// Writes variable-width value at specfied index into 'buffer'. Value must
+  /// not be null.
+  int32_t serializeVariableWidth(vector_size_t index, char* buffer);
+
+ private:
+  /// Returns serialized size of array row.
+  int32_t arrayRowSize(vector_size_t index);
+
+  /// Serializes array value to buffer. Value must not be null.
+  int32_t serializeArray(vector_size_t index, char* buffer);
+
+  /// Returns serialized size of map row.
+  int32_t mapRowSize(vector_size_t index);
+
+  /// Serializes map value to buffer. Value must not be null.
+  int32_t serializeMap(vector_size_t index, char* buffer);
+
+  /// Returns serialized size of a range of values.
+  int32_t arrayRowSize(
+      UnsafeRowFast& elements,
+      vector_size_t offset,
+      vector_size_t size,
+      bool fixedWidth);
+
+  /// Serializes a range of values into buffer using UnsafeRow Array
```

Comment:
Also comment
```
... The function returns the serialized array in bytes.
```

dittos for the other serialized methods. Thanks!

Reply:
@xiaoxmeng Just to clarify, are you asking to comment about the return value? This function doesn't return serialized array, rather the number of bytes written to 'buffer'.

Reply:
I meant the serialized array size in bytes